File: app/src/main/java/com/thumbtrek/app/ui/Charts.kt
package com.thumbtrek.app.ui

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.PathMeasure
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.thumbtrek.app.data.TRACKED_APPS
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/** One arc of [TrekGauge]. [key] should be the package name: it picks the colour. */
data class ChartSlice(
    val key: String,
    val label: String,
    val value: Float,
    val valueLabel: String,
)

/** One column of [HistoryBars]; [label] is the axis caption underneath it. */
data class ChartBar(val label: String, val value: Float)

/** One line of [TrendLines]. [key] should be the package name: it picks the colour. */
data class ChartSeries(val key: String, val label: String, val values: List<Float>)

/**
 * Categorical ramp for per-app series. Deliberately separate from the semantic accents in
 * [TrekPalette]: a chart hue means "which app", never "good" or "hot". Hues are spaced
 * around the wheel and held at similar lightness so no single app looks louder than the
 * others, and the light set is darkened rather than merely desaturated so it survives on
 * bone paper.
 */
private val SERIES_DARK = listOf(
    Color(0xFF8BE6A0),
    Color(0xFF77C4E8),
    Color(0xFFE9C063),
    Color(0xFFE79BB4),
    Color(0xFFB6A2F0),
    Color(0xFF6FD3C4),
)

private val SERIES_LIGHT = listOf(
    Color(0xFF1E6B3E),
    Color(0xFF1A5F87),
    Color(0xFF8A5806),
    Color(0xFF9C2F55),
    Color(0xFF5A3FA8),
    Color(0xFF116B60),
)

/**
 * Deterministic series colour: the same [key] always maps to the same hue, so an app keeps
 * its colour across every chart and the gauge, the split list and the trend all agree.
 * Tracked apps take their slot in [TRACKED_APPS] order; anything else hashes in.
 */
@Composable
fun chartColor(key: String): Color {
    val palette = if (Trek.isDark) SERIES_DARK else SERIES_LIGHT
    val index = TRACKED_APPS.keys.indexOf(key).takeIf { it >= 0 }
        ?: (key.hashCode() and Int.MAX_VALUE)
    return palette[index % palette.size]
}

// The gauge is an open-bottom instrument dial, not a closed donut: the break at the bottom
// is where the streak sits, and an arc that does not close reads as a measurement in
// progress rather than a completed pie.
private const val GAUGE_START = 130f
private const val GAUGE_SWEEP = 280f
private const val GAUGE_TICKS = 12

/**
 * PRD 5.2 per-app ring, drawn as a segmented survey dial. Sweeps open once on entry, then
 * animates between values. [content] is laid out in the middle, which is where the hero
 * distance lives on the dashboard.
 *
 * With no data the dial draws as a dashed, unwalked trail rather than disappearing, so a
 * first-run screen still shows the shape of the thing it is about to fill.
 */
@Composable
fun TrekGauge(
    slices: List<ChartSlice>,
    modifier: Modifier = Modifier,
    ringWidth: Dp = 14.dp,
    content: @Composable () -> Unit,
) {
    val drawn = slices.filter { it.value > 0f }
    val total = drawn.fold(0f) { acc, slice -> acc + slice.value }
    val colors = drawn.map { chartColor(it.key) }
    val track = Trek.groundSunken
    val tickColor = Trek.hairline
    val emptyColor = Trek.hairline
    val motion = LocalTrekMotion.current

    // Keyed on "is there anything to draw" rather than on the values themselves: the sweep
    // should play once, when the first real figures land, and never replay while today's
    // numbers tick up underneath it.
    val hasData = drawn.isNotEmpty() && total > 0f
    val reveal = remember { Animatable(0f) }
    LaunchedEffect(hasData, motion) {
        when {
            !hasData -> reveal.snapTo(0f)
            motion -> reveal.animateTo(1f, tween(TrekDur.REVEAL, easing = TrekEase))
            else -> reveal.snapTo(1f)
        }
    }

    Box(modifier = modifier, contentAlignment = Alignment.Center) {
        Canvas(
            modifier = Modifier
                .fillMaxSize()
                .clearAndSetSemantics { },
        ) {
            val stroke = ringWidth.toPx()
            val tickLength = stroke * 0.45f
            val tickGap = stroke * 0.75f
            val diameter = (size.minDimension - stroke - (tickLength + tickGap) * 2f)
                .coerceAtLeast(0f)
            val radius = diameter / 2f
            val cx = size.width / 2f
            val cy = size.height / 2f
            val topLeft = Offset(cx - radius, cy - radius)
            val arcSize = Size(diameter, diameter)

            // Survey ticks sit outside the ring, marking eighths of the dial.
            val tickRadius = radius + stroke / 2f + tickGap
            repeat(GAUGE_TICKS + 1) { i ->
                val angle = (GAUGE_START + GAUGE_SWEEP * i / GAUGE_TICKS) * PI.toFloat() / 180f
                val dx = cos(angle)
                val dy = sin(angle)
                val long = i % 3 == 0
                val len = if (long) tickLength else tickLength * 0.55f
                drawLine(
                    color = tickColor,
                    start = Offset(cx + dx * tickRadius, cy + dy * tickRadius),
                    end = Offset(cx + dx * (tickRadius + len), cy + dy * (tickRadius + len)),
                    strokeWidth = 1.5.dp.toPx(),
                    cap = StrokeCap.Round,
                )
            }

            if (drawn.isEmpty() || total <= 0f) {
                drawArc(
                    color = emptyColor,
                    startAngle = GAUGE_START,
                    sweepAngle = GAUGE_SWEEP,
                    useCenter = false,
                    topLeft = topLeft,
                    size = arcSize,
                    style = Stroke(
                        width = stroke * 0.5f,
                        cap = StrokeCap.Round,
                        pathEffect = PathEffect.dashPathEffect(
                            floatArrayOf(2.dp.toPx(), 7.dp.toPx()),
                        ),
                    ),
                )
                return@Canvas
            }

            drawArc(
                color = track,
                startAngle = GAUGE_START,
                sweepAngle = GAUGE_SWEEP,
                useCenter = false,
                topLeft = topLeft,
                size = arcSize,
                style = Stroke(width = stroke, cap = StrokeCap.Round),
            )

            val gap = if (drawn.size > 1) 2.5f else 0f
            var start = GAUGE_START
            drawn.forEachIndexed { index, slice ->
                val full = slice.value / total * GAUGE_SWEEP
                val sweep = (full - gap).coerceAtLeast(0.8f) * reveal.value
                drawArc(
                    color = colors[index],
                    startAngle = start + gap / 2f,
                    sweepAngle = sweep,
                    useCenter = false,
                    topLeft = topLeft,
                    size = arcSize,
                    style = Stroke(
                        width = stroke,
                        cap = if (drawn.size == 1) StrokeCap.Round else StrokeCap.Butt,
                    ),
                )
                start += full
            }
        }
        content()
    }
}

/**
 * PRD 5.3 history bars. One animation drives the whole set with a per-column offset, so a
 * long range still costs a single running animation instead of one per bar.
 *
 * @param bars ordered oldest to newest.
 * @param highlight index drawn in the accent colour, normally today. -1 for none.
 */
@Composable
fun HistoryBars(
    bars: List<ChartBar>,
    modifier: Modifier = Modifier,
    highlight: Int = -1,
    barColor: Color = Trek.slate,
    accentColor: Color = Trek.moss,
    height: Dp = 132.dp,
    maxInlineLabels: Int = 8,
) {
    if (bars.isEmpty()) return
    val max = bars.maxOf { it.value }.coerceAtLeast(1f)
    val track = Trek.groundSunken
    val motion = LocalTrekMotion.current

    // Same rule as the gauge: play once when data first exists, then hold. Keying on the
    // values would replay the whole sweep every time today's column ticks up.
    val hasData = bars.any { it.value > 0f }
    val grow = remember { Animatable(0f) }
    LaunchedEffect(bars.size, hasData, motion) {
        grow.snapTo(0f)
        if (hasData && motion) {
            grow.animateTo(1f, tween(TrekDur.LARGE + bars.size * 18, easing = TrekEase))
        } else {
            grow.snapTo(1f)
        }
    }

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Canvas(
            modifier = Modifier
                .fillMaxWidth()
                .height(height)
                .clearAndSetSemantics { },
        ) {
            val slot = size.width / bars.size
            val barWidth = (slot * 0.54f).coerceAtMost(22.dp.toPx()).coerceAtLeast(2f)
            val corner = CornerRadius(barWidth / 2f, barWidth / 2f)
            // Columns start together and finish in order, which reads as one gesture
            // sweeping left to right rather than a queue of separate animations.
            val span = 0.45f
            val stagger = if (bars.size > 1) (1f - span) / (bars.size - 1) else 0f

            bars.forEachIndexed { index, bar ->
                val x = slot * index + (slot - barWidth) / 2f
                drawRoundRect(
                    color = track,
                    topLeft = Offset(x, 0f),
                    size = Size(barWidth, size.height),
                    cornerRadius = corner,
                )
                if (bar.value <= 0f) return@forEachIndexed
                val local = ((grow.value - index * stagger) / span).coerceIn(0f, 1f)
                val target = (bar.value / max * size.height).coerceIn(barWidth, size.height)
                val barHeight = target * local
                if (barHeight <= 0f) return@forEachIndexed
                drawRoundRect(
                    color = if (index == highlight) accentColor else barColor,
                    topLeft = Offset(x, size.height - barHeight),
                    size = Size(barWidth, barHeight),
                    cornerRadius = corner,
                )
            }
        }
        ChartAxisLabels(bars.map { it.label }, maxInlineLabels, highlight)
    }
}

/**
 * PRD 5.3 per-app trends: one line per app over a shared x axis. The paths draw themselves
 * in with [PathMeasure], which is why a series appears as a stroke being laid down rather
 * than a shape fading up.
 */
@Composable
fun TrendLines(
    series: List<ChartSeries>,
    labels: List<String>,
    modifier: Modifier = Modifier,
    height: Dp = 128.dp,
    maxInlineLabels: Int = 8,
) {
    val drawn = series.filter { it.values.isNotEmpty() }
    if (drawn.isEmpty()) return
    val points = drawn.maxOf { it.values.size }
    val max = drawn.maxOf { line -> line.values.maxOrNull() ?: 0f }.coerceAtLeast(1f)
    val colors = drawn.map { chartColor(it.key) }
    val baseline = Trek.hairline
    val motion = LocalTrekMotion.current

    val trace = remember { Animatable(0f) }
    LaunchedEffect(drawn.size, points, motion) {
        trace.snapTo(0f)
        if (motion) {
            trace.animateTo(1f, tween(TrekDur.REVEAL, easing = TrekEase))
        } else {
            trace.snapTo(1f)
        }
    }

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Canvas(
            modifier = Modifier
                .fillMaxWidth()
                .height(height)
                .clearAndSetSemantics { },
        ) {
            val stroke = 2.dp.toPx()
            val inset = stroke * 2f
            val plotHeight = (size.height - inset * 2f).coerceAtLeast(1f)
            val plotWidth = (size.width - inset * 2f).coerceAtLeast(1f)
            val stepX = if (points > 1) plotWidth / (points - 1) else 0f

            drawLine(
                color = baseline,
                start = Offset(0f, size.height - inset),
                end = Offset(size.width, size.height - inset),
                strokeWidth = 1.dp.toPx(),
            )

            val measure = PathMeasure()
            drawn.forEachIndexed { lineIndex, line ->
                fun pointAt(index: Int): Offset {
                    val x = if (points > 1) inset + stepX * index else size.width / 2f
                    val value = line.values[index].coerceIn(0f, max)
                    return Offset(x, size.height - inset - value / max * plotHeight)
                }
                if (line.values.size == 1) {
                    drawCircle(
                        color = colors[lineIndex],
                        radius = stroke * 1.6f * trace.value,
                        center = pointAt(0),
                    )
                    return@forEachIndexed
                }

                val path = Path()
                line.values.indices.forEach { index ->
                    val point = pointAt(index)
                    if (index == 0) path.moveTo(point.x, point.y) else path.lineTo(point.x, point.y)
                }
                val visible = if (trace.value >= 1f) {
                    path
                } else {
                    measure.setPath(path, false)
                    Path().also { measure.getSegment(0f, measure.length * trace.value, it, true) }
                }
                drawPath(
                    path = visible,
                    color = colors[lineIndex],
                    style = Stroke(
                        width = stroke,
                        cap = StrokeCap.Round,
                        join = StrokeJoin.Round,
                    ),
                )
                if (trace.value >= 1f) {
                    drawCircle(
                        color = colors[lineIndex],
                        radius = stroke * 1.7f,
                        center = pointAt(line.values.lastIndex),
                    )
                }
            }
        }
        if (labels.isNotEmpty()) {
            ChartAxisLabels(labels, maxInlineLabels, highlight = labels.lastIndex)
        }
    }
}

/**
 * A seven-column glance strip: no axis, no frame, just the shape of the week with today
 * picked out. Small enough to sit under the hero without competing with it.
 */
@Composable
fun WeekPulse(
    values: List<Float>,
    modifier: Modifier = Modifier,
    barColor: Color = Trek.slate,
    accentColor: Color = Trek.moss,
    height: Dp = 34.dp,
) {
    if (values.isEmpty()) return
    val max = values.maxOrNull()?.coerceAtLeast(0.0001f) ?: return
    val track = Trek.groundSunken
    val motion = LocalTrekMotion.current
    val hasData = values.any { it > 0f }
    val grow = remember { Animatable(0f) }
    LaunchedEffect(values.size, hasData, motion) {
        grow.snapTo(0f)
        if (hasData && motion) {
            grow.animateTo(1f, tween(TrekDur.LARGE, easing = TrekEase))
        } else {
            grow.snapTo(1f)
        }
    }

    Canvas(
        modifier = modifier
            .fillMaxWidth()
            .height(height)
            .clearAndSetSemantics { },
    ) {
        val slot = size.width / values.size
        val barWidth = (slot * 0.46f).coerceAtMost(14.dp.toPx()).coerceAtLeast(2f)
        val corner = CornerRadius(barWidth / 2f, barWidth / 2f)
        values.forEachIndexed { index, value ->
            val x = slot * index + (slot - barWidth) / 2f
            drawRoundRect(
                color = track,
                topLeft = Offset(x, 0f),
                size = Size(barWidth, size.height),
                cornerRadius = corner,
            )
            if (value <= 0f) return@forEachIndexed
            val h = (value / max * size.height).coerceIn(barWidth, size.height) * grow.value
            drawRoundRect(
                color = if (index == values.lastIndex) accentColor else barColor,
                topLeft = Offset(x, size.height - h),
                size = Size(barWidth, h),
                cornerRadius = corner,
            )
        }
    }
}

/** Legend entry: colour key, series name, optional value. */
@Composable
fun ChartLegendRow(color: Color, label: String, value: String?, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        SeriesDot(color)
        Text(
            label,
            modifier = Modifier.weight(1f),
            style = MaterialTheme.typography.bodyMedium,
            color = Trek.ink,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
        if (value != null) {
            Text(value, style = TrekFigure, color = Trek.inkMuted, maxLines = 1)
        }
    }
}

@Composable
private fun ChartAxisLabels(labels: List<String>, maxInline: Int, highlight: Int) {
    if (labels.isEmpty()) return
    val muted = Trek.inkFaint
    val strong = Trek.ink
    if (labels.size <= maxInline) {
        Row(modifier = Modifier.fillMaxWidth()) {
            labels.forEachIndexed { index, label ->
                Text(
                    label,
                    modifier = Modifier.weight(1f),
                    style = TrekOverline,
                    color = if (index == highlight) strong else muted,
                    maxLines = 1,
                    overflow = TextOverflow.Clip,
                    textAlign = TextAlign.Center,
                )
            }
        }
    } else {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            Text(labels.first(), style = TrekOverline, color = muted, maxLines = 1)
            Text(labels.last(), style = TrekOverline, color = strong, maxLines = 1)
        }
    }
}

/** Ring of progress toward a locked achievement, drawn around its glyph. */
@Composable
fun ProgressRing(
    progress: Float,
    modifier: Modifier = Modifier,
    size: Dp = 44.dp,
    color: Color = Trek.moss,
    track: Color = Trek.hairline,
    stroke: Dp = 2.5.dp,
    content: @Composable () -> Unit,
) {
    val motion = LocalTrekMotion.current
    val swept = remember { Animatable(0f) }
    LaunchedEffect(progress, motion) {
        if (motion) {
            swept.animateTo(progress.coerceIn(0f, 1f), tween(TrekDur.LARGE, easing = TrekEase))
        } else {
            swept.snapTo(progress.coerceIn(0f, 1f))
        }
    }
    Box(
        modifier = modifier.size(size),
        contentAlignment = Alignment.Center,
    ) {
        Canvas(modifier = Modifier.fillMaxSize().clearAndSetSemantics { }) {
            val w = stroke.toPx()
            val inset = w / 2f
            val arcSize = Size(this.size.width - w, this.size.height - w)
            drawArc(
                color = track,
                startAngle = -90f,
                sweepAngle = 360f,
                useCenter = false,
                topLeft = Offset(inset, inset),
                size = arcSize,
                style = Stroke(width = w),
            )
            if (swept.value > 0f) {
                drawArc(
                    color = color,
                    startAngle = -90f,
                    sweepAngle = 360f * swept.value,
                    useCenter = false,
                    topLeft = Offset(inset, inset),
                    size = arcSize,
                    style = Stroke(width = w, cap = StrokeCap.Round),
                )
            }
        }
        content()
    }
}

/** Spacer that reserves a chart's height while its data is still loading. */
@Composable
fun ChartPlaceholder(height: Dp, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .height(height)
            .padding(vertical = 8.dp),
        verticalArrangement = Arrangement.Bottom,
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalAlignment = Alignment.Bottom,
        ) {
            listOf(0.4f, 0.7f, 0.3f, 0.9f, 0.55f, 0.75f, 0.45f).forEach { f ->
                Skeleton(
                    modifier = Modifier.weight(1f),
                    height = (height.value * f).dp,
                )
            }
        }
        Spacer(modifier = Modifier.height(8.dp))
    }
}

File: linux/thumbtrek/charts.py
"""Cairo chart widgets — geometry ported from app/.../ui/Charts.kt.

TrekGauge: open-bottom survey dial (start 130°, sweep 280°, 13 outside
ticks), per-app slices with 2.5° gaps, dashed unwalked-trail arc when empty.
HistoryBars: rounded columns on sunken tracks, slate, moss highlight, tap to
select. WeekPulse, TrendLines, ProgressRing, Rail likewise mirror their
Kotlin twins. No text is drawn here — axis labels are Gtk widgets.
"""

from __future__ import annotations

import math

import gi

gi.require_version("Gtk", "4.0")

from gi.repository import Gtk  # noqa: E402

from .theme import series_color  # noqa: E402

GAUGE_START = 130.0
GAUGE_SWEEP = 280.0
GAUGE_TICKS = 12


def _rgb(hex_color: str) -> tuple[float, float, float]:
    hex_color = hex_color.lstrip("#")
    return tuple(int(hex_color[i:i + 2], 16) / 255.0 for i in (0, 2, 4))


def _set(cr, hex_color: str, alpha: float = 1.0) -> None:
    cr.set_source_rgba(*_rgb(hex_color), alpha)


def _round_rect(cr, x: float, y: float, w: float, h: float, r: float) -> None:
    r = min(r, w / 2.0, h / 2.0)
    cr.new_path()
    cr.arc(x + w - r, y + r, r, -math.pi / 2, 0)
    cr.arc(x + w - r, y + h - r, r, 0, math.pi / 2)
    cr.arc(x + r, y + h - r, r, math.pi / 2, math.pi)
    cr.arc(x + r, y + r, r, math.pi, 3 * math.pi / 2)
    cr.close_path()


class Rail(Gtk.DrawingArea):
    """The one bar primitive: sunken track, filled portion, min cap radius."""

    def __init__(self, fraction: float = 0.0, color: str = "#1E6B3E",
                 track: str = "#E6E3D6", height: int = 6):
        super().__init__()
        self.fraction = fraction
        self.color = color
        self.track = track
        self.set_size_request(-1, height)
        self.set_draw_func(self._draw, None)

    def set(self, fraction: float, color: str | None = None) -> None:
        self.fraction = fraction
        if color is not None:
            self.color = color
        self.queue_draw()

    def _draw(self, _area, cr, width: int, height: int, _data) -> None:
        shown = min(1.0, max(0.0, self.fraction))
        _set(cr, self.track)
        _round_rect(cr, 0, 0, width, height, height / 2.0)
        cr.fill()
        if shown > 0:
            filled = max(width * shown, height)
            _set(cr, self.color)
            _round_rect(cr, 0, 0, filled, height, height / 2.0)
            cr.fill()


class Gauge(Gtk.DrawingArea):
    """Segmented survey dial with the hero content in the middle (TrekGauge)."""

    def __init__(self, dark: bool = False):
        super().__init__()
        self.dark = dark
        self.slices: list[tuple[str, float]] = []  # (app key, value)
        self.set_size_request(320, 320)
        self.set_draw_func(self._draw, None)

    def set(self, slices: list[tuple[str, float]], dark: bool) -> None:
        self.slices = [(k, v) for k, v in slices if v > 0]
        self.dark = dark
        self.queue_draw()

    def _draw(self, _area, cr, width: int, height: int, _data) -> None:
        from .theme import DARK, LIGHT
        p = DARK if self.dark else LIGHT
        drawn = self.slices
        total = sum(v for _, v in drawn)
        stroke = 15.0
        tick_len, tick_gap = stroke * 0.45, stroke * 0.75
        diameter = max(0.0, min(width, height) - stroke - (tick_len + tick_gap) * 2)
        radius = diameter / 2.0
        cx, cy = width / 2.0, height / 2.0

        tick_radius = radius + stroke / 2.0 + tick_gap
        _set(cr, p["hairline"])
        cr.set_line_width(1.5)
        cr.set_line_cap(1)  # round
        for i in range(GAUGE_TICKS + 1):
            angle = math.radians(GAUGE_START + GAUGE_SWEEP * i / GAUGE_TICKS)
            dx, dy = math.cos(angle), math.sin(angle)
            length = tick_len if i % 3 == 0 else tick_len * 0.55
            cr.move_to(cx + dx * tick_radius, cy + dy * tick_radius)
            cr.line_to(cx + dx * (tick_radius + length), cy + dy * (tick_radius + length))
            cr.stroke()

        start_rad = math.radians(GAUGE_START)
        sweep_rad = math.radians(GAUGE_SWEEP)
        if not drawn or total <= 0:
            _set(cr, p["hairline"])
            cr.set_line_width(stroke * 0.5)
            cr.set_line_cap(1)
            cr.set_dash([2.0, 7.0])
            cr.arc(cx, cy, radius, start_rad, start_rad + sweep_rad)
            cr.stroke()
            cr.set_dash([])
            return

        _set(cr, p["sunken"])
        cr.set_line_width(stroke)
        cr.set_line_cap(0)
        cr.arc(cx, cy, radius, start_rad, start_rad + sweep_rad)
        cr.stroke()

        gap = math.radians(2.5) if len(drawn) > 1 else 0.0
        angle = start_rad
        for index, (key, value) in enumerate(drawn):
            full = value / total * sweep_rad
            sweep = max(full - gap, math.radians(0.8))
            _set(cr, series_color(key, self.dark))
            cr.set_line_width(stroke)
            cr.set_line_cap(0 if len(drawn) > 1 else 1)  # butt, round when solo
            cr.arc(cx, cy, radius, angle + gap / 2.0, angle + gap / 2.0 + sweep)
            cr.stroke()
            angle += full


class HistoryBars(Gtk.DrawingArea):
    """Tappable history columns (HistoryBars): slate, moss highlight, ink pick."""

    def __init__(self, dark: bool = False, on_select=None):
        super().__init__()
        self.dark = dark
        self.values: list[float] = []
        self.highlight = -1
        self.selected = -1
        self.on_select = on_select
        self.set_size_request(-1, 132)
        self.set_draw_func(self._draw, None)
        gesture = Gtk.GestureClick()
        gesture.connect("pressed", self._pressed)
        self.add_controller(gesture)

    def set(self, values: list[float], highlight: int, selected: int, dark: bool) -> None:
        self.values, self.highlight, self.selected, self.dark = values, highlight, selected, dark
        self.queue_draw()

    def _pressed(self, _gesture, _n: int, x: float, _y: float) -> None:
        if not self.values or self.on_select is None:
            return
        width = self.get_allocated_width()
        index = min(int(x / width * len(self.values)), len(self.values) - 1)
        self.on_select(index)

    def _draw(self, _area, cr, width: int, height: int, _data) -> None:
        from .theme import DARK, LIGHT
        p = DARK if self.dark else LIGHT
        if not self.values:
            return
        peak = max(max(self.values), 1e-9)
        slot = width / len(self.values)
        bar_w = min(max(slot * 0.54, 2.0), 22.0)
        has_selection = 0 <= self.selected < len(self.values)
        for i, value in enumerate(self.values):
            x = slot * i + (slot - bar_w) / 2.0
            _set(cr, p["sunken"])
            _round_rect(cr, x, 0, bar_w, height, bar_w / 2.0)
            cr.fill()
            if value <= 0:
                continue
            bar_h = min(max(value / peak * height, bar_w), height)
            if i == self.selected:
                _set(cr, p["ink"])
            elif has_selection:
                _set(cr, p["slate"], 0.4)
            elif i == self.highlight:
                _set(cr, p["moss"])
            else:
                _set(cr, p["slate"])
            _round_rect(cr, x, height - bar_h, bar_w, bar_h, bar_w / 2.0)
            cr.fill()


class WeekPulse(Gtk.DrawingArea):
    """Seven-column glance strip; today in moss (WeekPulse)."""

    def __init__(self, dark: bool = False):
        super().__init__()
        self.dark = dark
        self.values: list[float] = []
        self.set_size_request(-1, 34)
        self.set_draw_func(self._draw, None)

    def set(self, values: list[float], dark: bool) -> None:
        self.values, self.dark = values, dark
        self.queue_draw()

    def _draw(self, _area, cr, width: int, height: int, _data) -> None:
        from .theme import DARK, LIGHT
        p = DARK if self.dark else LIGHT
        if not self.values:
            return
        peak = max(max(self.values), 1e-4)
        slot = width / len(self.values)
        bar_w = min(max(slot * 0.46, 2.0), 14.0)
        for i, value in enumerate(self.values):
            x = slot * i + (slot - bar_w) / 2.0
            _set(cr, p["sunken"])
            _round_rect(cr, x, 0, bar_w, height, bar_w / 2.0)
            cr.fill()
            if value <= 0:
                continue
            bar_h = min(max(value / peak * height, bar_w), height)
            _set(cr, p["moss"] if i == len(self.values) - 1 else p["slate"])
            _round_rect(cr, x, height - bar_h, bar_w, bar_h, bar_w / 2.0)
            cr.fill()


class TrendLines(Gtk.DrawingArea):
    """Per-app trend strokes with endpoint dots (TrendLines)."""

    def __init__(self, dark: bool = False):
        super().__init__()
        self.dark = dark
        self.series: list[tuple[str, list[float]]] = []
        self.set_size_request(-1, 128)
        self.set_draw_func(self._draw, None)

    def set(self, series: list[tuple[str, list[float]]], dark: bool) -> None:
        self.series = [(k, v) for k, v in series if v]
        self.dark = dark
        self.queue_draw()

    def _draw(self, _area, cr, width: int, height: int, _data) -> None:
        from .theme import DARK, LIGHT
        p = DARK if self.dark else LIGHT
        drawn = self.series
        if not drawn:
            return
        points = max(len(v) for _, v in drawn)
        peak = max(max(v) for _, v in drawn + [(" ", [1.0])])
        inset, stroke = 4.0, 2.0
        plot_w = max(1.0, width - inset * 2)
        plot_h = max(1.0, height - inset * 2)
        step = plot_w / (points - 1) if points > 1 else 0.0
        _set(cr, p["hairline"])
        cr.set_line_width(1.0)
        cr.move_to(0, height - inset)
        cr.line_to(width, height - inset)
        cr.stroke()
        cr.set_line_width(stroke)
        cr.set_line_cap(1)
        cr.set_line_join(1)
        for key, values in drawn:
            _set(cr, series_color(key, self.dark))
            pts = [(inset + step * i if points > 1 else width / 2.0,
                    height - inset - min(max(v, 0.0), peak) / peak * plot_h)
                   for i, v in enumerate(values)]
            if len(pts) == 1:
                cr.arc(pts[0][0], pts[0][1], stroke * 1.6, 0, 2 * math.pi)
                cr.fill()
                continue
            cr.move_to(*pts[0])
            for pt in pts[1:]:
                cr.line_to(*pt)
            cr.stroke()
            cr.arc(pts[-1][0], pts[-1][1], stroke * 1.7, 0, 2 * math.pi)
            cr.fill()


class ProgressRing(Gtk.DrawingArea):
    """Achievement progress ring around a glyph (ProgressRing)."""

    def __init__(self, progress: float = 0.0, color: str = "#1E6B3E",
                 track: str = "#DCD8C8", dark: bool = False):
        super().__init__()
        self.progress = progress
        self.color = color
        self.track = track
        self.dark = dark
        self.set_size_request(42, 42)
        self.set_draw_func(self._draw, None)

    def _draw(self, _area, cr, width: int, height: int, _data) -> None:
        from .theme import DARK, LIGHT
        p = DARK if self.dark else LIGHT
        cx, cy = width / 2.0, height / 2.0
        radius = min(width, height) / 2.0 - 2.0
        _set(cr, p["hairline"])
        cr.set_line_width(2.5)
        cr.arc(cx, cy, radius, 0, 2 * math.pi)
        cr.stroke()
        shown = min(1.0, max(0.0, self.progress))
        if shown > 0:
            _set(cr, self.color)
            cr.set_line_width(2.5)
            cr.set_line_cap(1)
            cr.arc(cx, cy, radius, -math.pi / 2, -math.pi / 2 + shown * 2 * math.pi)
            cr.stroke()
